### DISCUSSION TITLE
Information of data set that could be helpful for users

### DISCUSSION THREAD

**Discussion by Phoebecdy91** [zowe/zowe-chat#120] [Ideas]
Hi, we are looking for feedback for the attributes info of a data set and its member.
Would love to know what kind of info can be helpful for users:
<img width="725" alt="image" src="https://user-images.githubusercontent.com/23204849/187619298-671383e9-6e90-4d3e-aaa9-6131f9722b5d.png">
<img width="802" alt="image" src="https://user-images.githubusercontent.com/23204849/187619526-40e09403-752e-4e4c-bb05-a7310bd75c6c.png">

@Joe-Winchester maybe you will have some thoughts?

**Comment by Joe-Winchester** [comment]
Great question !    A couple of reference points,  first is the amount of raw data it's possible to get back from the API itself.   This limits the scope of what is possible to report.  Some things (such as the syntax being JCL) aren't actually possible to determine.  This is usually inferred by the word `JCL` being somewhere in the name, however that isn't always the case and varies from site to site.  For that reason I think `Syntax:` needs to be dropped as it can't be determined with accuracy.   

The most important information is usually the `Record Format`, however rather than showing `FB` this is short for `Fixed Block`.  Likewise `U` is short for `Undefined`.  It'd be great for us to find the full set and render them with their long description alongside their two character mnemonic.  This [web page](https://www.ibm.com/docs/en/zos-basic-skills?topic=set-data-record-formats) has a good list of descriptions.  

For a `FB` PDS the record length is very important.  In the sample above you show `80`, which typically lets a user know it's going to be holding a language like JCL, REXX or COBOL.  Along with the record format the size is also an important attribute to show.  

This is the full list of attribute available to use

```
{
    "items": [
        {
            "dsname": "WINCHJ.JCL",
            "blksz": "6160",
            "catnm": "CATALOG.T50031",
            "cdate": "2021/03/26",
            "dev": "3390",
            "dsntp": "PDS",
            "dsorg": "PO",
            "edate": "***None***",
            "extx": "1",
            "lrecl": "80",
            "migr": "NO",
            "mvol": "N",
            "ovf": "NO",
            "rdate": "2022/09/01",
            "recfm": "FB",
            "sizex": "15",
            "spacu": "CYLINDERS",
            "used": "60",
            "vol": "T50035",
            "vols": "T50035"
        }
    ],
}
```

Also, for reference this is a screen show of the same data set in ISPF 

<img width="1147" alt="image" src="https://user-images.githubusercontent.com/28100302/187911363-bdbfd7ea-3376-480a-930a-9cf1566f3731.png">

**Reply by FALLAI-Denis** [reply]
Hi @Joe-Winchester 

I think it is necessary to display the informations with the same values as those used by the Zowe CLI command line, this to avoid having a double naming plan and generating errors when using Zowe CLI commands.

Zowe CLI files use "--rf FB", so display "FB", not "Fixex Blocked".

**Comment by FALLAI-Denis** [comment]
Hi,

I think today sites work under control of DFHSMS (and DFHHSM) for managing user files.
Then in general, when working under DFHSMS control, many attributes are managed by DHFSMS and the attributes provided by the user are ignored or used to determine new values (role of ACS routines).

What seems important to me are the characteristics specific to the file or member of PDS: lrecl, blksize, allocated space, used space, allocated extends (limit of 16 extends by default). Information about storage devices is not important.

For the case of the members of PDS, it is necessary to distinguish the case of the loadlibs and the other PDS because the management of the statistics associated with the member is different.

For PDS : type of PDS, (PO=standard ou PO-E=library).
For standard PDS :  informations on Directory (x37 abend).
For loadlib PDS : informations on loadmodule from statistics, (AMODE/RMODE, alias. ...)
For non loadlib PDS : member statistics (last modifier, version/release, ...)

**Comment by Phoebecdy91** [comment]
I just start a co-creation session in Mural board:
https://app.mural.co/invitation/mural/zowecontainerizationresearch5083/1662625342433?sender=shdycai5839&key=9d640fb9-fe23-4ea1-a06c-d465d45b9ba7

Feel free to drag and drop and copy the sticky notes for the attributes prioritization and categorization.

Some tips for using Mural: 
> Move canvas: Click in blank area and drag or you could use the mini-map on the right bottom corner, then you could navigate to the other side of this canvas.
> Zoom in & out: Scroll up & down or you could use the mini-map on the right bottom corner.
> Create sticky notes: Double click anywhere, you could get one for yourself.
> Copy elements: Press "Option" for Mac user & "Alt" for Windows user and drag the element, you could directly copy one for yourself.

**Comment by Phoebecdy91** [comment]
This is based on the feedback today. Remove some attributes that might not be important for users.
**Need to confirm:** 
- does the attributes and priority make sense?
- How to get Free Space？ 
- What terms need to be more specific: like PO-E, 3390?
<img width="455" alt="图片" src="https://user-images.githubusercontent.com/23204849/190166104-6b6826d6-f1f5-44de-8f63-2d94b51ba083.png">
<img width="458" alt="图片" src="https://user-images.githubusercontent.com/23204849/190166138-61dbb5c3-84e7-447e-a8d2-34f11faae40a.png">

**Comment by Kevin-Miner** [comment]
As of z/OS 1.7:

• Basic SAM is 16 extents / data set
- Extended format is 123 extents / volume * 59 volumes / data set (Must be SMS man-aged
- Large Format Sam is 16 extents / volume * 59 volumes / data set
- Non-striped non-SMS VSAM is 255 / component 
- Non-striped SMS VSAM is 123 extents/volume * 59 volumes / component
- PDS is 16 extents per data set.  These data sets are restricted to a single volume, meaning they can not span volumes. 
- PDSE is 123 extents per data set. These data sets are restricted to a single volume, meaning they can not span volumes.
- Striped VSAM is 255 extents * number of stripes.  The maximum number of extents DADSM supports per volume is 123.
- A non-SMS VSAM is still limited to 255 extents per component.  Only SMS managed SMS with space constraint data class will you get 59*123 extents.

- In 1.7 the 255-extent limit for a VSAM component was removed. Prior to this sup-port, VSAM provided for a maximum of 255 extents per component for a non-striped data set, and 255 extents per stripe for a striped data set. 
- The system allows for 123 extents per volume * 59 volumes which is 7257 extents per component./stripe. The support applies to both VSAM and VSAM RLS, SMS managed data sets only.


- Sequential data sets are limited to 59 volumes.
- Basic Format:  SAM limit is 64K-1 tracks with half-track blocking that becomes (64K-1)*(27998)*2 bytes. This is 65,535 tracks, 4,369 cylinders, 3,669,697,860 bytes
- Large Format: has the following limits: With half-track blocking 2**24-1 tracks which is  (2**24 - 1)*(27998)*2 bytes. This is 16,777,215 tracks, 1118481 cylinders, 939,456,931,140 bytes, nearly 1TB of data.
- Extended Format: size limitations are 2**31-1 tracks, which is (2**31-1 ) * (27998-32)*2 bytes. This is 2,147,483,647 tracks, 143165576.46 cylinders, 120,113,055,344,004 bytes 

The standard sizes for z/OS DASD devices, also known as a volume, typically range from 1,113 cylinders to a maximum of 1,182,006 cylinders (17,730,090 tracks). 

As you can see the physical limitations for some data set types can be rather large.  Yet the limitations most data sets encounter are with the extent limitations which vary by data set type of either 16 or 123.

The system does not keep track of the primary space requested during the create process for any non-VSAM data set type, (basic, large, PDS or PDS/E).

Each time a new extent is needed to write to a data set the size of the extent requested is the secondary allocation unit.

The number of extents returned by the system for data set create or extend can vary be-tween 1 and 5 extents to satisfy the new space request.  For purposes of percentage of maximum data set size we will assume one extent is returned. 

When data is written to the data set, either during create or extend processing, the system keeps track of the actual space used in each extent.  For example if a data set is allocated SPACE=(CYL,(1,1) create or extend process will acquire one cylinder, 15 tracks, each time a new extent is needed.  The system tracks the amount of unused space in each extent.  For non-VSAM data sets this is in the Data Set Control Block (DSCB) in the VTOC.  For VSAM data sets this is in the VSAM Volume Record (VVR) in the VVDS.  

For purposes of the Zowe chat bot data set size limitations I recommend we focus on number of extents allocated to a data set and not the percentage of space used in each data set as this makes the calculations much easier to perform.  Simply it is the number of extents in a data set divided by the maximum number of extents allowed for the data set type.

**Comment by Phoebecdy91** [comment]
@Kevin-Miner Thanks Kevin for the detailed explanation.
Quick question for the data set member attributes:
- Do you know the full name of “cnorc”, “mnorc”, “inorc”, “sclm”?

<img width="600" alt="image" src="https://user-images.githubusercontent.com/23204849/191220548-a8e7ef52-afa3-4faa-8ae1-e67ebe1bd46e.png">

**Reply by Kevin-Miner** [reply]
SCLM is Software Configuration and Library Management.  The other fields appear to have be defined by zOSMF as I found this link, 
https://ibm.github.io/ibm_zos_zosmf/modules/zmf_dataset.html

**Reply by Phoebecdy91** [reply]
@Kevin-Miner THX! But it seems that there were no detailed full-name explanation of those attributes.
Is it something related to the "return code"?
<img width="707" alt="image" src="https://user-images.githubusercontent.com/23204849/191403698-68a0c091-7bb0-474b-8d48-d5b17ee6218a.png">

**Reply by FALLAI-Denis** [reply]
Perhaps here: https://www.ibm.com/docs/en/zos/2.5.0?topic=member-parameters

c4date: ZLC4DATE, Creation date, a character variable shown in your national format with a 4-character year. Use system variable ZDATEF to determine the national format. If no value exists for this variable, ISPF sets the value to blanks.

cnorc: ZLCNORC, Current number of records; a number in the range 0 - 65535. If no value exists for this variable, ISPF sets the value to blanks.

inorc: ZLINORC, Initial number of records; a number in the range 0 - 65535.

m4date: ZLM4DATE, Last change date, a character variable shown in your national format with a 4-character year. Use system variable ZDATEF to determine the national format. If no value exists for this variable, ISPF sets the value to blanks.

sclm; ZSCLM, Indicates whether the member was last modified by SCLM or ISPF. A value of Y indicates the last update was made through SCLM. A value of N indicates that the last update was made through ISPF.

**Reply by Phoebecdy91** [reply]
@FALLAI-Denis Thank you so much! This link is really helpful!
